File: clients/js/policy.js
// Ordered policy engine — first match wins; default applies when nothing matches.

export const ALLOW = "allow";
export const DENY = "deny";
export const APPROVE = "approve";

export class Decision {
  constructor(verdict, reason = "", rule = null) {
    this.verdict = verdict;
    this.reason = reason;
    this.rule = rule;
  }
  get allowed() { return this.verdict === ALLOW; }
  get denied() { return this.verdict === DENY; }
  get needsApproval() { return this.verdict === APPROVE; }
}

class Rule {
  constructor(verdict, name, when, reason) {
    this.verdict = verdict;
    this.name = name;
    this.when = when;
    this.reason = reason;
  }
  matches(action) {
    if (this.name != null && action.name !== this.name) return false;
    if (this.when != null && !this.when(action)) return false;
    return true;
  }
}

export class Policy {
  constructor(defaultVerdict = ALLOW) {
    if (![ALLOW, DENY, APPROVE].includes(defaultVerdict)) {
      throw new Error(`default must be one of allow/deny/approve, got ${defaultVerdict}`);
    }
    this.default = defaultVerdict;
    this.rules = [];
  }
  _add(verdict, name, when, reason) {
    this.rules.push(new Rule(verdict, name ?? null, when ?? null, reason || `matched ${verdict} rule`));
    return this;
  }
  allow(name, when, reason = "") { return this._add(ALLOW, name, when, reason); }
  deny(name, when, reason = "") { return this._add(DENY, name, when, reason); }
  requireApproval(name, when, reason = "") { return this._add(APPROVE, name, when, reason); }
  decide(action) {
    for (const rule of this.rules) {
      if (rule.matches(action)) return new Decision(rule.verdict, rule.reason, rule);
    }
    return new Decision(this.default, `default:${this.default}`);
  }
}

File: warrant/policy.py
"""A tiny, ordered policy engine: decide allow / deny / approve for an action before it runs.

Rules are evaluated in the order they were added; the FIRST match wins. If nothing
matches, the policy's `default` applies. Default is `allow` (frictionless for dev); set
`Policy(default="deny")` for a deny-by-default production posture.
"""

from __future__ import annotations

from typing import Callable, Optional

ALLOW, DENY, APPROVE = "allow", "deny", "approve"


class Decision:
    def __init__(self, verdict: str, reason: str = "", rule: "Optional[_Rule]" = None):
        self.verdict = verdict
        self.reason = reason
        self.rule = rule

    @property
    def allowed(self) -> bool:
        return self.verdict == ALLOW

    @property
    def denied(self) -> bool:
        return self.verdict == DENY

    @property
    def needs_approval(self) -> bool:
        return self.verdict == APPROVE

    def __repr__(self) -> str:
        return f"Decision({self.verdict!r}, reason={self.reason!r})"


class _Rule:
    def __init__(self, verdict: str, name: Optional[str],
                 when: Optional[Callable[[dict], bool]], reason: str):
        self.verdict, self.name, self.when, self.reason = verdict, name, when, reason

    def matches(self, action: dict) -> bool:
        if self.name is not None and action.get("name") != self.name:
            return False
        if self.when is not None and not self.when(action):
            return False
        return True


class Policy:
    def __init__(self, default: str = ALLOW):
        if default not in (ALLOW, DENY, APPROVE):
            raise ValueError(f"default must be one of allow/deny/approve, got {default!r}")
        self.default = default
        self.rules: list = []

    def _add(self, verdict, name, when, reason) -> "Policy":
        self.rules.append(_Rule(verdict, name, when, reason or f"matched {verdict} rule"))
        return self  # chainable

    def allow(self, name=None, when=None, reason="") -> "Policy":
        return self._add(ALLOW, name, when, reason)

    def deny(self, name=None, when=None, reason="") -> "Policy":
        return self._add(DENY, name, when, reason)

    def require_approval(self, name=None, when=None, reason="") -> "Policy":
        return self._add(APPROVE, name, when, reason)

    def decide(self, action: dict) -> Decision:
        for rule in self.rules:
            if rule.matches(action):
                return Decision(rule.verdict, rule.reason, rule)
        return Decision(self.default, f"default:{self.default}")
